### PR TITLE
String Interpolation

### PR DESCRIPTION
**Pull request by Licenser**
[Rendered](https://github.com/tremor-rs/tremor-rfcs/blob/string-interpolation/text/0000-string-interpolation.md)

### CODE REVIEW

**Comment by mfelsche** [comment]
Great RFC!

Unfortunately I think the current situation doesnt leave too much wiggle room for alternatives here:

A string like:

```
string::format("this should create a {{}} and say {}", 42) == "this should create a {} and say 42";
```

cannot really work right now. At runtime, we want to represent the string as compact as possible, a nice little continuous string would the the best option for performance reasons. In the resulting formatted string, we intend `{{}}` to be `{}` in the final result, it will be `\{\}` in the runtime representation, so that `string::format` doesnt use it as placeholder. When calling `string::format` we get our literal `{}` in the resulting formatted string. But what if we do not use string format? We only get a `\{\}` in the resulting string. What if we use interpolation? `"{ 42 } this should work and create a {{}} and say {}"` does create a runtime about mismatching count of args and placeholders, error even before we fully executed it.

Lets look at this:

```
string::format("this { "should work and" } create a {{}} and say {}", 42)
```

What this transforms into in the AST is:

```
string::format(string::format("this {} create a \{\} and say {}", "should work and"), 42)
```

This will obviously fail, as there is only additional argument to the inner `string::format` although the string somehow produced 2 placeholders. But to the user, this behaviour is very much not obvious.

To make this work as intended the string would need to look like:

```
string::format("this { "should work and" } create a {{{{}}}} and say {{}}", 42) == "this should work and create a {} and say 42"
```

Whereas when no interpolation is used, the above raises a runtime error as the number of placeholders and arguments does not match. It needs to be written as:

```
string::format("this { "should work and" } create a {{}} and say {}", 42) == "this should work and create a {} and say 42"
```

Having to use different escaping levels depending on whether interpolation is used or not, is not acceptable.

I think the only way to overcome this, is to use option 2. or 3. in the RFC.
While Option 4 seems nice, removing `string::format` means people cannot do purely dynamic formatting anymore as this cannot be replaced by string interpolation:

```
string::format(event.format_string, event.args)
```

Thus, i would vote for adding a prefix like `${}` (or any other) to interpolations.

**Comment by Licenser** [comment]
Very good point. And I think you're right. I expanded on the prior art section and added a conclusion section to reflect that input.

File: .github/pull_request_template.md [hunk 1]
```diff
@@ -0,0 +1 @@
+[Rendered](https://github.com/tremor-rs/tremor-rfcs/blob/<BRANCH>/text/<FILE>.md)
```
Comment (mfelsche) [inline .github/pull_request_template.md:1]: 👍

File: text/0000-string-interpolation.md [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+downsides:
+
+1. it is a breaking change (however every change would be)
+
+From the survey of of existing implementations the most common form seems to be `${...}`, followed closely by `#{...}` both seem workable however, both `$` and `#` have their own meaning in tremor script.
+
+- `#` is the comment character, so it can be reasond that it is switching contexts from code to comment, or from string to interpolation.
+- `$` is our `metadata`, which could be reasond that `${...}` is a meta context, however this seems to be a bit of a stretch and `${$}` would read quite odd.
+
+Considering all this, the proposed solution for this RFC is to use `#{...}` for string interpolation.
```
Comment (mfelsche) [inline text/0000-string-interpolation.md:190]: :+1:

File: .github/pull_request_template.md (hunk 1 above)
Comment (Licenser) [reply .github/pull_request_template.md:1]: :D only copyig from the best

**Review by mfelsche — APPROVED** [review]

**Pull request merged**